3_parameter_efficient_finetuning/lora_adapters.md

```diff
@@ -1,31 +1,61 @@
 # LoRA (Low-Rank Adaptation)
 
-LoRA has become the most widely adopted PEFT method. It works by adding small rank decomposition matrices to the attention weights, typically reducing trainable parameters by about 90%. Here's a basic configuration:
+LoRA has become the most widely adopted PEFT method. It works by adding small rank decomposition matrices to the attention weights, typically reducing trainable parameters by about 90%. 
 
-```python
-from peft import LoraConfig
+## Understanding LoRA
 
-config = LoraConfig(
-    task_type="CAUSAL_LM",
-    r=8,  # Rank of update matrices
-    lora_alpha=32,  # Scale of updates
-    lora_dropout=0.1
-)
-```
+LoRA (Low-Rank Adaptation) is a parameter-efficient fine-tuning technique that freezes the pre-trained model weights and injects trainable rank decomposition matrices into the model's layers. Instead of training all model parameters during fine-tuning, LoRA decomposes the weight updates into smaller matrices through low-rank decomposition, significantly reducing the number of trainable parameters while maintaining model performance. For example, when applied to GPT-3 175B, LoRA reduced trainable parameters by 10,000x and GPU memory requirements by 3x compared to full fine-tuning. You can read more about LoRA in the [LoRA paper](https://arxiv.org/pdf/2106.09685).
 
-This configuration can be used with TRL to train a model with LoRA. LoRA matrices are known as adapters and can be merged back into the base model.
+LoRA works by adding pairs of rank decomposition matrices to transformer layers, typically focusing on attention weights. During inference, these adapter weights can be merged with the base model, resulting in no additional latency overhead. LoRA is particularly useful for adapting large language models to specific tasks or domains while keeping resource requirements manageable.
 
 ## Merging LoRA Adapters
 
 After training with LoRA, you might want to merge the adapter weights back into the base model for easier deployment. This creates a single model with the combined weights, eliminating the need to load adapters separately during inference.
```

Comment:
I would first introduce code to simply load one or multiple adapters instead of focusing on the merging already. Are there clear benefits of merging besides the mentioned "easier than loading separately"? If so, can you mention them?

Reply:
At smaller model sizes like this, merging is more likely. But you're right, we could expand this for educational purposes. I've added this issue: https://github.com/huggingface/smol-course/issues/72#issue-2727951665